threescale/src/proxy.rs

```diff
@@ -92,15 +96,79 @@ pub fn set_app_id_to_cache(user_key: &UserKey, app_id: &AppId) -> Result<(), Cac
-        Some(cas),
-    ) {
-        Ok(()) => true,
-        Err(e) => {
-            debug!("set operation failed for key: {} : {:?}", key, e);
-            false
+    let prev_memory_usage = get_cache_pair_size(key) as i32;
+    let serialized_app = bincode::serialize::<Application>(app).unwrap();
+    let memory_delta: i32 = (key.len() as i32) + (serialized_app.len() as i32) - prev_memory_usage;
+    if let Err(e) = set_shared_data(key, Some(&serialized_app), Some(cas)) {
+        debug!("set operation failed for key: {} : {:?}", key, e);
+        return false;
+    }
+    for num_try in 0..3 {
+        match update_shared_memory_size(memory_delta) {
+            Ok(()) => break,
+            Err(e) => debug!("try#{} : failed to update memory counter: {}", num_try, e),
         }
     }
+    true
+}
+
+// returns memory used in bytes for both key and value pair stored.
+fn get_cache_pair_size(key: &str) -> usize {
+    match get_shared_data(key) {
+        Ok((Some(bytes), _)) => key.len() + bytes.len(),
+        Ok((None, _)) => key.len(),
+        Err(_) => 0,
+    }
+}
+
+// Adds delta bytes to the shared memory usage counter
+fn update_shared_memory_size(delta: i32) -> Result<(), anyhow::Error> {
+    let current_size = match get_shared_data(SHARED_MEMORY_KEY) {
+        Ok((Some(bytes), _)) => {
+            let arr: [u8; 8] = match bytes.try_into() {
+                Ok(res) => res,
+                Err(e) => anyhow::bail!("failed to convert vec<u8> to [u8;8]: {:?}", e),
+            };
+            u64::from_be_bytes(arr)
+        }
+        Ok((None, _)) => {
+            warn!("shared memory size was not initialized at the start or got deleted somehow!");
+            if let Err(e) = set_shared_data(
+                SHARED_MEMORY_KEY,
+                Some(&SHARED_MEMORY_INITIAL_SIZE.to_be_bytes()),
+                None,
+            ) {
+                anyhow::bail!(
+                    "failed to initialize shared memory size: {:?}",
+                    CacheError::ProxyStatus(e as u8)
+                )
+            }
+            SHARED_MEMORY_INITIAL_SIZE
+        }
+        Err(e) => anyhow::bail!(
+            "getting shared memory size failed: {:?}",
+            CacheError::ProxyStatus(e as u8)
+        ),
+    };
+
+    let final_size: u64;
+    if delta.is_negative() {
+        if current_size <= (-delta).try_into().unwrap() {
+            // This condition is theoretically not possible because memory should keep on
+            // increasing since we have no option of deleting the data. So, check your calcs!
+            final_size = SHARED_MEMORY_INITIAL_SIZE;
+        } else {
+            final_size = current_size.saturating_sub((-delta).try_into().unwrap());
+        }
+    } else {
+        final_size = current_size.saturating_add((-delta).try_into().unwrap());
+    }
+
+    if let Err(e) = set_shared_data(SHARED_MEMORY_KEY, Some(&final_size.to_be_bytes()), None) {
```

Comment:
You'll probably want to get the `cas` from `get_shared_data` above and reuse it here - if the call failed _with a CasMismatch_ then someone else updated this before you did, so... probably you don't need to care about it and can assume it was properly updated?

Reply:
Used CAS from above.